### PR TITLE
KAFKA-13817 Always sync nextTimeToEmit with wall clock

### PR DESCRIPTION
**Pull request by qingwei91**
We should sync nextTimeToEmit with wall clock on each method call to ensure throttling works correctly in case of clock drift.
If we dont, then in the event of significant clock drift, throttling might not happen for a long time, this can hurt performance.

I've added a unit test to simulate clock drift and verify my change works.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java
Comment (qingwei91) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:232]: Is it ok to have comments here? it wasn't obvious to me what this piece of code was doing initially, I thought having comments might help, but I don't feel strongly, please let me know if you'd like it removed

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java
Comment (qingwei91) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:361]: This test is quite convoluted because it relies on low-level API, this appears to be the 1st instance in test (other test relies on higher level API), is this acceptable?

I resort to this approach because we need to manipulate TimeTracker which isn't available in high level API. And I don't feel comfortable to make larger change in the codebase.

Please let me know if you think there's a better way.

Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:361]: It'a a minor improvement to the actual code, so it might be ok to not add a test for it? -- Otherwise, I don't have a proposal for better code... It's in the guts so it's messy (and thus maybe not worth?) to test.

Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:361]: There is `KStreamWindowAggregateTest#shouldEmitWithLargeInterval()` that tests a similar thing.

Comment (qingwei91) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:361]: I am happy to defer to you or fellow contributors/maintainers.

My personal view is that the behavior change is quite subtle, so having test to codify it is useful, but if we are happy to merge it without unit test I am happy too

**Comment by lihaosky** [comment]
I can also take a look by end of this week.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java
Comment (lihaosky) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:361]: Hi @qingwei91 , thanks for fixing and great test coverage! Regarding test complexity, can you do something similar as https://github.com/apache/kafka/blob/trunk/streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java#L768 to test time drift. Instead of mocking low level stores, can you check the final results?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:232]: I'm ok with the comment

**Review by lihaosky — COMMENTED** [review]
Thanks @qingwei91 ! The fix looks good. Just one comment about the test to see if we can make it simpler.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java
Comment (qingwei91) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:361]: Thanks for the advice, I will try to mimick that

**Comment by mjsax** [comment]
@qingwei91 -- What is the status of this PR? Seems there is open comments that would need to be addressed? Would be great if we could push this over the finish line.

**Comment by qingwei91** [comment]
@mjsax sorry, I will try to pick this back up this weekend

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java
Comment (qingwei91) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:361]: Hi @lihaosky , I changed the test to this: https://github.com/apache/kafka/pull/12166/commits/26f6fa6f809851c8e57c660472bcce47f881e536

Is it ok?

**Review by lihaosky — APPROVED** [review]
Thanks @qingwei91 ! LGTM. Failed test doesn't seem related.

**Comment by lihaosky** [comment]
@mjsax can help approve and merge as a committer.

**Comment by mjsax** [comment]
Thanks for the PR! Merged to `trunk`.

**Pull request merged**
